### ISSUE TITLE
Cannot create nodes with beta.kubernetes.io/fluentd-ds-ready=true label

### ISSUE THREAD

**Issue by underyx** [weaveworks/eksctl#1373]
**What happened?**
`eksctl create nodegroup` returns

```
[✖]  unknown 'kubernetes.io' or 'k8s.io' labels were specified: [beta.kubernetes.io/fluentd-ds-ready]
```

**What you expected to happen?**

Just let me use the label :confused: 

**How to reproduce it?**

Try to create a node group with 
```
    labels: { "beta.kubernetes.io/fluentd-ds-ready": "true" }
```

**Anything else we need to know?**

**Versions**
Please paste in the output of these commands:
```
$ eksctl version
[ℹ]  version.Info{BuiltAt:"", GitCommit:"", GitTag:"0.6.0"}
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:13:54Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.10-eks-5ac0f1", GitCommit:"5ac0f1d9ab2c254ea2b0ce3534fd72932094c6e1", GitTreeState:"clean", BuildDate:"2019-08-20T22:39:46Z", GoVersion:"go1.11.13", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by jrodguitar** [comment]
same here. just trying to specify some labels which are specified in a helm chart. and getting the same error.
unknown 'kubernetes.io' or 'k8s.io' labels were specified: [app.kubernetes.io/component app.kubernetes.io/instance app.kubernetes.io/name]

i tried in eksctl version 0.5.2 and 0.6.0

**Comment by blacabexg** [comment]
+1

**Comment by siarhei-karanets-epam** [comment]
i tried in eksctl version 0.15.0, still there is a problem.

**Comment by jmoscato** [comment]
This appears to be the relevant section of code for eksctl 0.17.0, in the ValidateNodeGroupLabels() function:
[link](https://github.com/weaveworks/eksctl/blob/0.17.0/pkg/apis/eksctl.io/v1alpha5/validation.go#L280)

```
if isKubernetesLabel && !allowedKubeletNamespace {
	switch l {
	case
		"kubernetes.io/hostname",
		"kubernetes.io/instance-type",
		"kubernetes.io/os",
		"kubernetes.io/arch",
		"beta.kubernetes.io/instance-type",
		"beta.kubernetes.io/os",
		"beta.kubernetes.io/arch",
		"failure-domain.beta.kubernetes.io/zone",
		"failure-domain.beta.kubernetes.io/region",
		"failure-domain.kubernetes.io/zone",
		"failure-domain.kubernetes.io/region":
	default:
		unknownKubernetesLabels = append(unknownKubernetesLabels, l)
	}
}
```

I don't see a way to override label validation or add to the list of allowed labels. Any chance something like that could be added as a feature? I'd like to avoid workarounds if possible (such as a daemonset that runs a job to label all nodes, externally labeling all nodes with kubectl, etc).

**Comment by Himangini** [comment]
find a best way to handle this

**Comment by Callisto13** [comment]
This is still present on `0.68.0-rc`.

But it is deliberate. Early 2019 eksctl code was changed to match k8s' behaviour of disallowing node labels which did not contain an approved prefix: https://github.com/weaveworks/eksctl/pull/495. It is likely that, even if `eksctl` _did_ allow labels with other prefixes, the kubelet itself would then error (at a later stage and in a much more annoying way).

I will look into how kubernetes behaves now and whether there is a way to get custom labels on nodes.

**Comment by Callisto13** [comment]
I was correct and this is by design.

Labels set in eksctl config are passed to the `--node-labels` kubelet flag in the `/etc/eks/bootstrap.sh` script. That kubelet flag is for [WellKnownLabels](https://kubernetes.io/docs/reference/labels-annotations-taints/) to be set at **start time**, and therefore they must meet a very specific pattern to not break things as the node joins the cluster.

The command `kubectl label nodes` is for adding extra and truly custom labels after start.

Ideally kubernetes would have called that flag `--register-with-labels` to make it very clear. Also `eksctl` could have picked a different and more descriptive name.

For now users will have to continue adding their labels in the same way that those not using eksctl would do: running the `kubectl` command after start.

On eksctl's side, we can make this easier.
Options:
- A) Instead of erroring when we collect a bunch of labels which are not WellKnown, and therefore not allowed at start, we can take that list and then set them via the k8s api after the nodes have started in a post create action.
- B) We create a new field in config, `extraLabels` or `customLabels` or smth, and after node creation we apply these via the k8s api in a post create action. Docs would need to clearly explain the difference

**Comment by Skarlso** [comment]
I would go with option B because that is more explicit and less of a hidden feature.

**Comment by TBBle** [comment]
The Well-Known Labels which make sense for the kubelet to self-apply are already covered [as exceptions in the `IsKubeletLabel` logic](https://github.com/kubernetes/kubernetes/blob/v1.22.2/staging/src/k8s.io/kubelet/pkg/apis/well_known_labels.go#L35-L53).

Generally, it's a bad practice for people to apply arbitrary labels from a namespace they don't own.

In this particular ticket, the user wanted to add a label which was removed and deprecated as part of the KEP I linked earlier, and which is not a current Well-Known Label, but for legacy reasons might be useful, if someone is still using the fluentd-elasticsearch addon configured [before kubernetes 1.14](https://github.com/kubernetes/kubernetes/pull/74424) and not updated since. If the user was attempting to create another 1.13 cluster, and got this failure that'd be an eksctl bug, as the label restriction did not apply there, and a few of the kubernetes addons depended on such node labels for DaemonSet functionality in 1.13.

This ticket has also been marked as a duplicate of one of the tickets about `node-role.kubernetes.io/*` support, which is similarly not a Well-Known Label, but is used by kubectl (at least) for display purposes.

In summary, this use-case is about _not_ Well-Known Labels in the Kubernetes-owned namespaces.

With the removal or renaming of the daemonset labels from the addons in 1.14, the only one I'm aware of that's neither "years-old legacy" or "a bad idea, don't do that" is `node-role.kubernetes.io/*`, and [that one's specifically called-out as restricted](https://github.com/kubernetes/kubernetes/issues/84912#issuecomment-551362981).

**Comment by Skarlso** [comment]
Sure, and that's what the PR is about, however, these labels will disappear the moment the node is rotated out because of AWS or other reasons. They will never be reapplied because they have been applied with the kube SDK and not through launch template or tags or the stack. So I'm struggling to see the point of this ticket for now.

**Comment by TBBle** [comment]
I somewhat agree, I don't think the proposed fix is long-term workable for those reasons. I think the fix for applying restricted labels will probably end up in the cloud provider implementation (https://github.com/weaveworks/eksctl/issues/2363#issuecomment-830745792) and possibly Managed Node Groups will have a different solution, as described at https://github.com/weaveworks/eksctl/issues/4007#issuecomment-886046421. 

Personally, I'd mark this ticket as a duplicate of #2363, since #2363 was already being used to aggregate the other reports coming from the same change for unmanaged clusters, and the specific label from this ticket stopped being relevant in the two years since it was opened, but #2363 revolves around a label that still has value today.
